Automattic/wp-calypso
# Checkout: Add list of discounts to sidebar

## Proposed Changes

This PR adds a list of all checkout discounts to the sidebar, grouped by type below a new item that reads "Subtotal before discounts". The list includes one item for each type of discount no matter how many items in the cart have this discount (eg: if a coupon applies to two items in the cart, it will show only one coupon discount in the sidebar which includes the sum of both; the same is true if two items have a first year free or a prorated discount).

Some common discounts (their labels originate on the server via the "cost overrides" system and each one also has a unique key that can be used to identify it in case we want to do something special later) are as follows:

- `Coupon code`
- `Item on sale`
- `Prorated balance from previous plan`
- `Free domain for first year` 

This should help users better understand what discounts are being applied rather than relying on all the details on each line item (which can get really messy in some cases). It also helps to set the stage for the v2 checkout redesign (https://github.com/Automattic/payments-shilling/issues/1969) which already has this list of discounts in the sidebar. 

> [!NOTE]
> Introductory offer discounts are a little tricky. Currently, they are not actually "discounts" in the way that our payments system considers them. Rather, they are a unique thing wherein the discounted offer price actually becomes the original price of the product. This gets complicated depending on which original cost you're talking about because once the original cost reaches the cart response, it has changed to be the undiscounted price (see https://github.com/Automattic/payments-shilling/issues/2341#issuecomment-1899060553 for a breakdown of how this works). A more consistent solution is to treat introductory offers as discounts, which I'm working on in D134600-code but before then, I've implemented a workaround in this PR to display them as if they were discounts. In either case, introductory offer discounts will be listed as "Introductory offer", which is not very explicit (it does not explain the products which have offers, nor does it explain the length of these offers (1 year? 3 months? etc.) We have that information available and so the discount item could be modified in a future PR to display a more nuanced view but doing so may be complicated to get right. For now it should be ok since the details are available in other parts of the checkout UI - the sidebar list is just a summary.

> [!NOTE]
> Sale discounts will only be displayed as "Item on sale", which is not very explicit as they don't include the items that are on sale. This will be improved by https://github.com/Automattic/wp-calypso/pull/86371 after this PR is merged but isn't included here because there's some complexity to get it looking right. For now this should be ok since the details are available in other parts of the checkout UI - the sidebar is just a summary.

Before (sidebar)            |  After (sidebar)
:-------------------------:|:-------------------------:
<img width="530" alt="Screenshot 2024-01-12 at 2 10 47 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/9c893e2e-8c42-4092-8e77-7c93c4c0ce90"> | <img width="528" alt="Screenshot 2024-01-12 at 7 46 30 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/8937eb61-1ba6-4853-8cd9-140acc39405c">

Before (introductory offer)            |  After (introductory offer)
:-------------------------:|:-------------------------:
<img width="726" alt="Screenshot 2024-01-17 at 7 59 12 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/2f3ffd22-4813-414f-8a91-568119659ed4"> | <img width="734" alt="Screenshot 2024-01-17 at 7 48 14 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/f2d618bf-20a6-437b-8042-18742c2a10c4">

Before  (w/ credits)           |  After (w/ credits)
:-------------------------:|:-------------------------:
<img width="232" alt="Screenshot 2024-01-12 at 5 55 51 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/b0b00019-5c91-48c4-9e93-ee7c1a697de6"> | <img width="238" alt="Screenshot 2024-01-12 at 5 54 36 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/4db76a05-74a9-4739-a104-3e600ee3f5e0">

Before (footer)           |  After (footer)
:-------------------------:|:-------------------------:
<img width="298" alt="Screenshot 2024-01-12 at 7 47 45 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/c6406f18-8470-4219-ac3b-48411b9295fe"> | <img width="302" alt="Screenshot 2024-01-12 at 7 47 36 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/84732664-a9f3-4e84-83b2-8d24989a2b0b">

## Testing Instructions

Add products to your cart with cost overrides (sale coupons, coupon codes, bundled domain, etc.).

Visit checkout and verify that you see the cost overrides in the sidebar and that they look good.

Also test with credits!
